src/web_socket.toit

```diff
@@ -0,0 +1,405 @@
+    headers.add "Connection" "upgrade"
+    headers.add "Upgrade" "websocket"
+    headers.add "Sec-WebSocket-Key" nonce
+    headers.add "Sec-WebSocket-Version" "13"
+    return nonce
+
+  static check_client_upgrade_response_ response/Response nonce/string [on_error] -> none:
+    if response.status_code != STATUS_SWITCHING_PROTOCOLS:
+      on_error.call response
+      throw response.stringify
+    upgrade_header := response.headers.single "Upgrade"
+    connection_header := response.headers.single "Connection"
+    if not upgrade_header
+        or not connection_header
+        or (Headers.ascii_normalize_ upgrade_header) != "Websocket"
+        or (Headers.ascii_normalize_ connection_header) != "Upgrade"
+        or (response.headers.single "Sec-WebSocket-Accept") != (WebSocket.response_ nonce):
+      throw "MISSING_HEADER_IN_RESPONSE"
+    if response.headers.single "Sec-WebSocket-Extensions"
+        or response.headers.single "Sec-WebSocket-Protocol":
+      throw "UNKNOWN_HEADER_IN_RESPONSE"
+
+  /**
+  Checks whether the request is a WebSocket upgrade request.
+  If it is a valid upgrade request, adds the required headers to the response_writer
+    and sends a response confirming the upgrade.
+  Otherwise responds with an error code and returns null.
+  */
+  static check_server_upgrade_request_ request/Request response_writer/ResponseWriter_ -> string?:
+    connection_header := request.headers.single "Connection"
+    upgrade_header := request.headers.single "Upgrade"
+    version_header := request.headers.single "Sec-WebSocket-Version"
+    nonce := request.headers.single "Sec-WebSocket-Key"
+    message := null
+    if nonce == null:                message="No nonce"
+    else if nonce.size != 24:        message="Bad nonce size"
+    else if connection_header != "upgrade": message="No Connection: upgrade"
+    else if upgrade_header != "websocket":  message="No Upgrade: websocket"
+    else if version_header != "13":         message="Unrecognized Websocket version"
+    else:
+      response_writer.headers.add "Sec-WebSocket-Accept" (response_ nonce)
+      response_writer.headers.add "Connection" "upgrade"
+      response_writer.headers.add "Upgrade" "websocket"
+      return nonce
+    response_writer.write_headers STATUS_BAD_REQUEST --message=message
+    return null
+
+  static response_ nonce/string -> string:
+    expected_response := base64.encode
+        sha1 nonce + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
+    return expected_response
+
+/**
+A writer for writing a single message on a WebSocket connection.
+*/
+class WebSocketWriter:
+  owner_ /WebSocket? := ?
+  size_ /int?
+  remaining_in_fragment_ /int := 0
+  fragment_sent_ /bool := false
+
+  constructor.private_ .owner_ .size_:
+
+  write data from=0 to=data.size -> int:
+    if owner_ == null: throw "ALREADY_CLOSED"
+    total_size := to - from
+    while true:
+      while from < to and remaining_in_fragment_ != 0:
```

Comment:
Reorder this function.
The header is sent first, so it should be on the top of the while loop.
It's really confusing to first see the code for the body, and then only the header.